INTERNAL ERROR when enabling mypy-zope for Twisted Klein

In commit https://github.com/twisted/klein/pull/471/commits/04a536c8b228d687fadd0d3e325753914d3e3b85, I enabled `mypy-zope` for [Twisted Klein](https://github.com/twisted/klein) and I'm [getting an INTERNAL ERROR in the build](https://github.com/twisted/klein/runs/2090195305).

Here is the error with `--show-traceback` added:

```console
wsanchez$ tox -e mypy -- --show-traceback src
GLOB sdist-make: /Users/wsanchez/Dropbox/Developer/Twisted/klein/setup.py
mypy inst-nodeps: /Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/.tmp/package/1/klein-20.6.0.zip
mypy installed: attrs==20.3.0,Automat==20.2.0,characteristic==14.3.0,constantly==15.1.0,hyperlink==21.0.0,idna==3.1,incremental==17.5.0,klein @ file:///Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/.tmp/package/1/klein-20.6.0.zip,mypy==0.812,mypy-extensions==0.4.3,mypy-zope==0.2.11,PyHamcrest==2.0.2,six==1.15.0,Tubes==0.2.0,Twisted==20.3.0,typed-ast==1.4.2,typing-extensions==3.7.4.3,Werkzeug==1.0.1,zope.event==4.5.0,zope.interface==5.2.0,zope.schema==6.1.0
mypy run-test-pre: PYTHONHASHSEED='3790822665'
mypy run-test: commands[0] | mypy --cache-dir=/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy_cache --pretty --show-traceback src
src/klein/test/test_trial.py:33: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.rtfd.io/en/latest/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.812
Traceback (most recent call last):
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/bin/mypy", line 8, in <module>
    sys.exit(console_entry())
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/__main__.py", line 11, in console_entry
    main(None, sys.stdout, sys.stderr)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/main.py", line 90, in main
    res = build.build(sources, options, None, flush_errors, fscache, stdout, stderr)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/build.py", line 179, in build
    result = _build(
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/build.py", line 253, in _build
    graph = dispatch(sources, manager, stdout)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/build.py", line 2638, in dispatch
    process_graph(graph, manager)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/build.py", line 2962, in process_graph
    process_stale_scc(graph, scc, manager)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/build.py", line 3060, in process_stale_scc
    graph[id].type_check_first_pass()
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/build.py", line 2130, in type_check_first_pass
    self.type_checker().check_first_pass()
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/checker.py", line 294, in check_first_pass
    self.accept(d)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/checker.py", line 401, in accept
    stmt.accept(self)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/nodes.py", line 950, in accept
    return visitor.visit_class_def(self)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/checker.py", line 1723, in visit_class_def
    self.accept(defn.defs)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/checker.py", line 401, in accept
    stmt.accept(self)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/nodes.py", line 1015, in accept
    return visitor.visit_block(self)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/checker.py", line 1972, in visit_block
    self.accept(s)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/checker.py", line 401, in accept
    stmt.accept(self)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/nodes.py", line 950, in accept
    return visitor.visit_class_def(self)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/checker.py", line 1749, in visit_class_def
    sig, _ = self.expr_checker.check_call(dec, [temp],
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/checkexpr.py", line 930, in check_call
    result = self.check_call(call_function, args, arg_kinds, context, arg_names,
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/checkexpr.py", line 911, in check_call
    return self.check_callable_call(callee, args, arg_kinds, context, arg_names,
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/checkexpr.py", line 1023, in check_callable_call
    new_ret_type = self.apply_function_plugin(
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy/checkexpr.py", line 732, in apply_function_plugin
    return method_callback(
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy_zope/plugin.py", line 171, in analyze_implementation
    iface_type = self._lookup_type(ifacename, method_ctx.api)
  File "/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/lib/python3.9/site-packages/mypy_zope/plugin.py", line 583, in _lookup_type
    module = api.modules[module_name]
KeyError: 'klein.test.test_trial.TestCaseTests'
src/klein/test/test_trial.py:33: : note: use --pdb to drop into pdb
ERROR: InvocationError for command /Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy/bin/mypy --cache-dir=/Users/wsanchez/Dropbox/Developer/Twisted/klein/.tox/mypy_cache --pretty --show-traceback src (exited with code 2)
__________________________________________ summary __________________________________________
ERROR:   mypy: commands failed
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,7 +7,7 @@ All notable changes to this project will be documented in this file.
 
 - Better type inference for contextmanagers and open() function: do not override 
   the default behavior for tese cases
-
+- Support interfaces defined inside classes
 
 0.2.11 (2021-02-21)
 -------------------

--- a/src/mypy_zope/plugin.py
+++ b/src/mypy_zope/plugin.py
@@ -1,5 +1,5 @@
 import sys
-from typing import List, Dict, Any, Callable, Optional
+from typing import List, Dict, Any, Callable, Optional, Tuple, Sequence, Union
 from typing import Type as PyType
 from typing import cast
 
@@ -51,6 +51,8 @@ from mypy.nodes import (
     ARG_OPT,
     FUNC_NO_INFO,
     NameExpr,
+    MypyFile,
+    SymbolNode,
 )
 
 from collections import defaultdict
@@ -585,17 +587,30 @@ class ZopeInterfacePlugin(Plugin):
                 pass
 
     def _lookup_type(self, fullname: str, api: TypeChecker) -> Instance:
-        # Implement own type lookup, because TypeChecker.named_generic_type is
-        # dysfunctional
-        parts = fullname.split(".")
-        module_name = ".".join(parts[:-1])
-        type_name = parts[-1]
-        module = api.modules[module_name]
-        sym = module.names[type_name]
-        typeinfo = sym.node
+        module, names = self._find_module(fullname, api)
+        container: Union[MypyFile, TypeInfo] = module
+        for name in names:
+            sym = container.names[name]
+            assert isinstance(sym.node, TypeInfo)
+            container = sym.node
+        typeinfo = container
         assert isinstance(typeinfo, TypeInfo)
+
         any_type = AnyType(TypeOfAny.from_omitted_generics)
         return Instance(typeinfo, [any_type] * len(typeinfo.defn.type_vars))
+
+    def _find_module(
+        self, fullname: str, api: TypeChecker
+    ) -> Tuple[MypyFile, Sequence[str]]:
+        parts = fullname.split(".")
+        for pn in range(len(parts) - 1, 0, -1):
+            moduleparts = parts[:pn]
+            names = parts[pn:]
+            modulename = ".".join(moduleparts)
+            if modulename in api.modules:
+                return api.modules[modulename], names
+
+        raise LookupError(fullname)
 
     def _index_members(self, typeinfo: TypeInfo) -> Dict[str, TypeInfo]:
         # we skip "object" and "Interface" since everyone implements it

--- a/tests/samples/nested_definitions.py
+++ b/tests/samples/nested_definitions.py
@@ -1,0 +1,22 @@
+import zope.interface
+
+
+class Host:
+    class IBookmark(zope.interface.Interface):
+        def get_url() -> str:
+            pass
+
+    @zope.interface.implementer(IBookmark)
+    class Bookmark:
+        def get_url(self) -> str:
+            return "http://"
+
+
+bm: Host.IBookmark = Host.Bookmark()
+reveal_type(bm)
+
+"""
+<output>
+nested_definitions.py:16: note: Revealed type is '__main__.Host.IBookmark'
+</output>
+"""

